onnxruntime/core/providers/webnn/builders/impl/qdq_op_builder.cc

```diff
@@ -0,0 +1,152 @@
+  int32_t zero_point_type = 0;
+  ORT_RETURN_IF_NOT(GetType(*input_defs[0], input_type, logger), "Cannot get input data type");
+  ORT_RETURN_IF_NOT(GetType(*output_defs[0], output_type, logger), "Cannot get output data type");
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val scale = model_builder.GetOperand(input_defs[1]->Name());
+
+  emscripten::val zero_point = emscripten::val::null();
+  if (input_defs.size() == 3 && input_defs[2]->Exists()) {
+    zero_point = model_builder.GetOperand(node.InputDefs()[2]->Name());
+  } else {
+    // DequantizeLinear: x_zero_point's data type equals to input data type
+    // QuantizeLinear: x_zero_point's data type equals to output data type
+    zero_point_type = op_type == "DequantizeLinear" ? input_type : output_type;
+    zero_point = model_builder.GetZeroConstant(zero_point_type);
+  }
+
+  emscripten::val output;
+  NodeAttrHelper helper(node);
+  int32_t axis = helper.Get("axis", 1);
+  int32_t block_size = helper.Get("block_size", 0);
+  // axis is valid for input shape greater than 1D.
+  if (input_shape.size() > 1) {
+    axis = static_cast<int32_t>(HandleNegativeAxis(axis, input_shape.size()));
+  }
+  // Insert ones before and after the axis dimension for broadcasting of 1D scale tensor.
+  if (1 == scale_shape.size() && 1 < input_shape.size()) {
+    std::vector<int32_t> target_shape{static_cast<int>(input_shape[axis])};
+    target_shape.insert(target_shape.begin(), axis, 1);
+    target_shape.insert(target_shape.end(), input_shape.size() - axis - 1, 1);
+    emscripten::val reshape_scale_options = emscripten::val::object();
+    reshape_scale_options.set("label", node.Name() + "_reshape_scale");
+    scale = model_builder.GetBuilder().call<emscripten::val>("reshape",
+                                                             scale,
+                                                             emscripten::val::array(target_shape),
+                                                             reshape_scale_options);
+    emscripten::val reshape_zero_point_options = emscripten::val::object();
+    reshape_zero_point_options.set("label", node.Name() + "_reshape_zero_point");
+    zero_point = model_builder.GetBuilder().call<emscripten::val>("reshape",
+                                                                  zero_point,
+                                                                  emscripten::val::array(target_shape),
+                                                                  reshape_zero_point_options);
+  }
+
+  // If block_size is specified, we need to expand the scale and zero_point tensors.
+  if (block_size > 1) {
+    emscripten::val concat_scale_inputs = emscripten::val::array();
+    emscripten::val concat_zero_point_inputs = emscripten::val::array();
+    for (int i = 0; i < block_size; i++) {
+      concat_scale_inputs.call<void>("push", scale);
+      concat_zero_point_inputs.call<void>("push", zero_point);
```

Comment:
> A `reshape` combined with `tile` or `expand` would be much faster than multiple concatenations. e.g. pseudocode
> 
> ```
> reshapedScale = reshape(scale, [scale.dimensions[0],1]) // Enable broadcast expansion.
> expandedScale = expand(reshapedScale, [scale.dimensions[0], blockSize]) // Expand to block size.
> expandedScale = reshape(expandedScale , [scale.dimensions[0] * blockSize]) // Flatten back to 1D.
> ```

For blocked dequantization, scale has the same shape as the input. If we reshape it to 1D, does WebNN support broadcasting it?

Reply:
> FYI. I'm thinking _ultimately_ we'll want to send this through to WebNN _without_ expansion, permitting WebNN's `dequantizeLinear` to accept scale and zero-point tensors that are not necessarily unidirectionally broadcastable, but they are an integer multiple (e.g. an input tensor of sizes = [128] and scale tensor of sizes = [32] where each scale element is shared by 4 input elements), because otherwise we lose memory saving benefits. I know DML 6.3 supports this...
> 
> ```
> enum DML_QUANTIZATION_TYPE
> {
>     DML_QUANTIZATION_TYPE_NONE,
>     DML_QUANTIZATION_TYPE_SCALE,
>     DML_QUANTIZATION_TYPE_SCALE_ZERO_POINT,
> };
> ...
> // The scale and zero point in QuantizationTensors should be a multiple of the input size (if not, the input left
> struct DML_DEQUANTIZE_OPERATOR_DESC
> {
>     const DML_TENSOR_DESC* InputTensor;
>     DML_QUANTIZATION_TYPE QuantizationType;
>     UINT QuantizationTensorCount;
>     _Field_size_(QuantizationTensorCount) const DML_TENSOR_DESC* QuantizationTensors;
>     const DML_TENSOR_DESC* OutputTensor;
> };
> ```
> 
> ...but I don't anticipate CoreML and TFLite do. Needs more research... 🤔 Our findings on Phi-3 will help determine this.

👍，it'd be great if the WebNN spec could handle the expansion.

In Phi-3 model, all DequantizeLinear nodes contain the `block_size` attribute:
![image](https://github.com/user-attachments/assets/ac4a47a9-dd3c-4d25-bfc0-c47bc7672c11)

![image](https://github.com/user-attachments/assets/53f14ab6-7740-4c3c-8d87-5c5a895884cc)

Reply:
> A `reshape` combined with `tile` or `expand` would be much faster than multiple concatenations. e.g. pseudocode
> 
> ```
> reshapedScale = reshape(scale, [scale.dimensions[0],1]) // Enable broadcast expansion.
> expandedScale = expand(reshapedScale, [scale.dimensions[0], blockSize]) // Expand to block size.
> expandedScale = reshape(expandedScale , [scale.dimensions[0] * blockSize]) // Flatten back to 1D.
> ```

@fdwr, finally, I choose `tile` and fix at https://github.com/microsoft/onnxruntime/pull/22180/commits/30cb16561fe49ecf1661745e70f8f6063271a714, PTAL, thanks!

Reply:
I don't think concatenating the scale / zeroPoint tensors `block_size` times along `axis` is correct. 

According to ONNX doc: https://onnx.ai/onnx/operators/onnx__DequantizeLinear.html
> block_size - INT (default is '0'):
> The size of the quantization block (number of times every scale is replicated)

Within a quantization block, the scale or zeroPoint value should be the same. This is as known as block-wise quantization.

So, IIUC, we should slice each scale / zeorPoint out and repeat it `block_size` times along `axis`. After all the expansions done, concatenate them all together along `axis`. The pseudo code for above phi3 case would be:

```js
scale0 =  slice(scale, [0, 0], [1, 9216]);
scaleForBlock0 = expand(scale0, [128, 9126]);
...
scale23 = slice(scale, [22, 0], [1, 9126]);
scaleForBlock23 = expand(scale23, [128, 9126]);

expandedScale = concat(scaleForBlock0, ..., scaleForBlock23); // shape [3072, 9216]
```

Reply:
Thanks @huningxin, I've verified that this is correct behavior, as I can get same result as wasm EP.

Besides, this also introduced a number of additional nodes into the graph, memory overhead increases about 8GB -> 10GB. (compared with `tile`)